### ISSUE TITLE
EPIC: v2.0 AST-Native Kuzu Backend

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#592]
# EPIC: Migrate to AST-Native Graph Storage

## Overview

Replace text-based Kuzu index with AST-native graph storage. Currently we store text and re-parse 3× per query (M1, reranker, extraction). New design stores parsed ASTs as native graph nodes, eliminating parsing overhead.

**Performance impact:** 100-1000× speedup (zero parsing at query time)

## Current Architecture (v1.0)

```
Indexing:  Text → Parse to AST → Serialize to text → Store in Kuzu
Querying:  Retrieve text → Parse (M1) → Parse (rerank) → Parse (extract)
           ^^^^^^^^^^^^^^   WASTEFUL - 3× parsing per query!
```

## Target Architecture (v2.0)

```
Indexing:  Text → Parse to AST → Store AST as graph nodes → DONE
Querying:  Retrieve AST nodes directly → Use immediately (ZERO parsing!)
```

### Document Hierarchy
- `SourceCollection` → `Document` → `Section` → `Paragraph` → `Sentence`
- Ordering edges: `NEXT_SECTION`, `NEXT_PARAGRAPH`, `NEXT_SENTENCE`
- Enables: Pronoun resolution, context retrieval, topic-based search

### AST Storage (Immutable)
- `AST` node: Versioned parser output (fraztipo, demandotipo, parse_statistics)
- `Frazo`, `Vortgrupo`, `Vorto` nodes: Full syntax tree
- All parser fields captured: affixes, correlatives, compounds, parse errors
- Versioning: Multiple AST versions per sentence (parser improvements)

### Extensible Annotations (Mutable)
- `AnnotationType` + `AnnotationValue` + `Annotation` pattern
- Zero schema changes for new annotation types
- Annotations at any level (Word, Sentence, Paragraph, Section, Document)
- Annotation vocabulary links to Root nodes (self-describing!)

### Key Innovation: Circular Vocabulary
- Annotation values ARE Esperanto words
- `AnnotationValue("PERSON")` → links to `Root("person")`
- Expand searches using synonyms: "person" → "homo", "individuo"
- The annotation vocabulary IS the graph vocabulary!

### ✅ No Model Retraining Required
- Models work on ASTs/embeddings, not storage format
- Compositional embeddings: No change
- Reranker: No change
- M1: No change

### ✅ Scripts Requiring Rewrites
1. **Indexing** (`scripts/index_kuzu.py`): Complete rewrite
   - Parse once, store AST nodes
   - Build document hierarchy
   - Create ordering edges
   
2. **Retrieval** (`klareco/rag/kuzu_inverted_index.py`): Major changes
   - Query graph structure directly
   - Return AST nodes (not text)
   - Zero parsing overhead

3. **Evaluation** (`scripts/evaluate_rag_test_set.py`): Minor changes
   - Receive pre-parsed ASTs
   - Remove triple-parsing code
   
4. **Answer Extraction** (`klareco/rag/answer_extractor.py`): Minor changes
   - Receive AST from retrieval
   - No parsing needed

### ❌ No Changes Needed
- Parser (`klareco/parser.py`): No changes
- Deparser (`klareco/deparser.py`): No changes
- Models (`klareco/models/`, `klareco/embeddings/`): No changes
- Training scripts: No changes

### Option A: New Branch + v2.0 (RECOMMENDED ✅)
- Create `v2.0-ast-native` branch
- Fresh start: Port only what we want to keep
- Leave behind accumulated cruft
- Merge back when stable

**Advantages:**
- Clean codebase
- No legacy compatibility burden
- Can refactor aggressively
- v1.0-snapshot tag preserves current state

**Disadvantages:**
- Temporary code duplication
- Requires full system testing before merge

### Option B: Incremental Migration (NOT RECOMMENDED ❌)
- Dual indexes (old + new)
- Feature flags to switch between backends
- Gradual migration of components

**Disadvantages:**
- Complexity explosion
- Maintains cruft
- Longer migration period
- Technical debt accumulation

### Phase 1: Core Schema (Week 1-2)
- [ ] Create `v2.0-ast-native` branch from `v1.0-snapshot`
- [ ] Implement core node tables: AST, Frazo, Vortgrupo, Vorto
- [ ] Implement document hierarchy: Document, Section, Paragraph, Sentence
- [ ] Write schema creation script
- [ ] Write schema validation tests

### Phase 2: Indexing (Week 2-3)
- [ ] Rewrite `scripts/index_kuzu_v2.py`
- [ ] Implement AST → graph conversion
- [ ] Implement document hierarchy extraction
- [ ] Add progress tracking + checkpointing
- [ ] Index small corpus subset (1000 sentences)
- [ ] Verify schema correctness

### Phase 3: Basic Retrieval (Week 3-4)
- [ ] Rewrite `klareco/rag/kuzu_inverted_index_v2.py`
- [ ] Implement root-based retrieval (BM25 equivalent)
- [ ] Return AST nodes directly
- [ ] Test on 10 sample queries
- [ ] Benchmark: Compare parsing time (should be ~0ms)

### Phase 4: Semantic Constraints (Week 4-5)
- [ ] Port semantic constraints (#585-#589) to graph queries
- [ ] Implement role-based filtering (subject/object)
- [ ] Implement entity boosting (WHO questions)
- [ ] Implement semantic similarity using graph traversal
- [ ] Test on general knowledge test set

### Phase 5: Full Pipeline Integration (Week 5-6)
- [ ] Update answer extractor to receive ASTs
- [ ] Update evaluation script
- [ ] Remove all parsing code from query path
- [ ] Run full evaluation on 30-question test set
- [ ] Compare performance: v1.0 vs v2.0

### Phase 6: Annotation System (Week 6-7)
- [ ] Implement extensible annotation schema
- [ ] Port entity classifier annotations
- [ ] Implement AnnotationType + AnnotationValue pattern
- [ ] Add annotation versioning
- [ ] Test annotation queries

### Phase 7: Advanced Features (Week 7-8)
- [ ] Implement sentence ordering (pronoun resolution support)
- [ ] Implement topic annotations
- [ ] Add Wikipedia category integration
- [ ] Test multi-level queries (section → paragraph → sentence)

### Phase 8: Full Corpus Indexing (Week 8-9)
- [ ] Index full Wikipedia corpus (~5M sentences)
- [ ] Monitor memory usage
- [ ] Verify graph integrity
- [ ] Benchmark query performance

### Phase 9: Testing + Stabilization (Week 9-10)
- [ ] Full test suite
- [ ] Performance benchmarks
- [ ] Documentation updates
- [ ] Migration guide

### Phase 10: Merge to Master (Week 10)
- [ ] Code review
- [ ] Merge `v2.0-ast-native` → `master`
- [ ] Tag as `v2.0.0`
- [ ] Archive v1.0 code

### Performance
- Query time: <50ms per question (vs ~2min currently)
- Zero parsing overhead during queries
- Memory: <4GB RAM for 5M sentences

### Functionality
- General knowledge test accuracy: ≥60% (baseline from v1.0)
- AST reconstruction: 100% fidelity
- Annotation extensibility: Add new type without schema change

### Code Quality
- Test coverage: ≥85% for new code
- Clean architecture: No v1.0 cruft
- Documentation: Complete API docs

## Migration Checklist

Before starting:
- [x] Create `v1.0-snapshot` tag
- [x] Push all commits to origin
- [ ] Create `v2.0-ast-native` branch
- [ ] Document v1.0 limitations (for future reference)

During development:
- [ ] Keep v1.0 branch stable (no new features)
- [ ] Regular merges from master (bug fixes only)
- [ ] Maintain parallel test suites

## Risks + Mitigations

**Risk 1: Kuzu schema complexity**
- Mitigation: Comprehensive schema validation tests
- Mitigation: Visual schema diagrams

**Risk 2: Graph query performance**
- Mitigation: Early benchmarking
- Mitigation: Index optimization

**Risk 3: Memory usage with full corpus**
- Mitigation: Incremental indexing with monitoring
- Mitigation: Kuzu is disk-backed (like SQLite)

**Risk 4: Merge conflicts when rebasing**
- Mitigation: Keep v1.0 stable (no new features)
- Mitigation: Regular small merges

## Open Questions

1. Should we version ASTs at indexing time or lazily on re-parse?
2. How to handle annotation conflicts (multiple systems annotating same word)?
3. Should we implement annotation rollback/undo?
4. What's the retention policy for old AST versions?

## Related Issues

- #585, #586, #587, #589: Semantic constraints (implemented in v1.0)
- Performance issue: Triple-parsing per query

## References

- Design discussion: [This conversation]
- v1.0 snapshot: `v1.0-snapshot` tag
- Schema diagrams: TBD

---

**Recommendation: Proceed with Option A (new branch, clean break)**

This is a significant enough change that we should start fresh, port what we want, and leave behind accumulated technical debt. The v1.0-snapshot tag preserves the current working state.
